### ISSUE TITLE
move volume definition from compose to image

### ISSUE THREAD

**Issue by nimbius** [netbox-community/netbox-docker#242]
Can the volumes used by netbox be defined in the image instead of compose?  it would improve portability and cross-platform operability outside the docker ecosystem.

## Desired Behavior

podman and docker run would see volumes created.  

## Contrast to Current Behavior

compose or an image built using 'from netboxcommunity/netbox' is required to get volumes outside compose.

## Changes Required

OCIv1 should have simple stanzas for the defined volumes please :)

## Discussion: Benefits and Drawbacks

reducing vendor-specific dependencies in the container helps improve its portability and and introduction into container orchestrators with a minimum of file conversion or rewriting.  closer adherence to the OCI standard helps eschew systems that promote vendor lock-in.

- What are the drawbacks of this change? Is it backwards-compatible?
docker-compose yaml may conflict or cause problems with image defined volumes.  It will require some testing or maybe a new tag release specific to compose based deployments.

- Anything else that you think is relevant to the discussion of this feature/change request.

**Comment by cimnine** [comment]
Thank you for starting this discussion.

Forgive me if I don't understand your request, but can you please explain why you believe it is necessary (or why it would be better) to define the volumes in the `Dockerfile`? What part of OCI mandates this?

There are two reasons why I'm asking:

1. As far as I understand all OCI runtimes are capable of mounting volumes to random paths without a `VOLUME` being defined in the Dockerfile. Docker (our target runtime) certainly is, and given that k8s is able to do it cri-o seems to be able to do this as well.
2. Volumes are – strictly speaking – not even required to use the image. I.e. the image should start fine without any volumes mounted. The configuration can be fed through environment variables and any files that are uploaded would not be persisted across restarts. All static files could be extracted beforehand and served by a completely independent HTTP server, e.g. though a multi-step build.

Not using docker-compose should therefore be perfectly fine. And there are projects out there which use our image without docker-compose, but with a different container management system, such as Kubernetes.

One could also start all the services using plain old `docker` commands and do the plumbing themselves. It's just harder. We use `docker-compose` because it's the simplest way to get this image up and running on a single machine with all the necessary dependencies pre-configured.

tl;dr: We aim for a "pull and ~pray~play" experience (that's `docker-compose` part) without actively hindering any alternative container management solution.

**Comment by nimbius** [comment]
I think 'changes required' should have been better defined on my part. There is no express requirement in the OCI to define volumes at any level, at all.  The concern is with the docker-compose file.  Unless lines 13-20 are transposed into the containers image definition, theyll need to be specified at runtime or converted to a kubernetes pod definition.  Both are easy enough to do, but trip up more advanced deployments outside the docker ecosystem..

My argument I suppose targets more advanced deployments.  while one can specify env based configuration, its certainly not ideal.  volumes in the docker compose file would need to be transcribed either to runtime -v specifications, or buildah would need to create a new image from the dockerhub image that specifis volumes in lines 13-18.  Volumes certainly arent required for netbox to run, but if you're targeting a larger or more production deployment, you want to make sure features like initializers and persistent media arent something you need to add later. 

the downside to this change is that it might make netbox less pull and play if compose defined volumes somehow conflict identically defined image volumes, but i dont have compose to test with.  Its entirely possible compose could inherit the volume definitions from the container image perhaps?

another alternative to this change is just submitting better documentation for the volume paths.

**Comment by cimnine** [comment]
> There is no express requirement in the OCI to define volumes at any level, at all.

I've just re-read the thread and I still don't see the necessity. With any container tooling that I know it is possible to mount arbitrary volumes and sometimes files into any container. There is no requirement for a `VOLUME` in the Dockerfile for this.

To me it even strikes me as a negative thing to have the `VOLUME` in the Dockerfile: At least Docker will then automatically mount an empty-dir, and this is even contra-productive. Because this will simply overwrite what is already baked into the image. (E.g. the `startup_scripts` are baked into the image and are only explicitly mounted in the `docker-compose.yml` file for convenience.)

> while one can specify env based configuration, its certainly not ideal.

That's why one can either overwrite the `configuration.py` file via a mount or secret file – or use this project's configuration file and leverage secret files just for certain values as described [in the wiki section called _Deployment_](https://github.com/netbox-community/netbox-docker/wiki/Deployment).

I have also discussed this with the other maintainer, Tobias. He was not convinced that this change is necessary, either.

So I will close this issue know. If you or someone else can provide a concrete case where the `VOLUME`s are really necessary because there is no reasonable way to make that case work otherwise I'm willing to re-consider.
